### PR TITLE
Enhance monitor graph params parser

### PR DESCRIPTION
**Pull request by aiwantaozi**
Problem:
Need to enhance monitor graph params parser
Solution:
Enhance params parser and add validation test

### CODE REVIEW

File: pkg/api/customization/monitor/handler_utils.go [hunk 1]
```diff
@@ -63,40 +65,102 @@ func (p *clusterGraphInputParser) parseFilter() error {
...
+
+	if p.ResourceType == "" {
+		p.ResourceType = p.Input.Filters["displayResourceType"]
+	}
+
+	if p.ResourceType == "" {
+		return fmt.Errorf("resourceType is empty")
+	}
+
+	clusterResourceTypes := []string{
```
Comment (orangedeng) [inline pkg/api/customization/monitor/handler_utils.go]: Maybe we can use `map[string]struct{}` instead of `[]string`.

File: pkg/api/customization/monitor/handler_utils.go [hunk 2]
```diff
@@ -113,34 +177,108 @@ func (p *projectGraphInputParser) parseFilter() error {
 		return fmt.Errorf("clusterName is empty")
 	}
 
+	p.ResourceType = p.Input.Filters["resourceType"]
+	if p.ResourceType == "" {
+		return fmt.Errorf("resourceType is empty")
+	}
+
+	projectResourceTypes := []string{
```
Comment (orangedeng) [inline pkg/api/customization/monitor/handler_utils.go]: Maybe we can use `map[string]struct{}` instead of `[]string`.

File: pkg/api/customization/monitor/handler_utils.go [hunk 3]
```diff
@@ -63,40 +65,102 @@ func (p *clusterGraphInputParser) parseFilter() error {
...
+		instance := p.Input.MetricParams["instance"]
+		if instance == "" {
+			return fmt.Errorf("instance in metric params is empty")
+		}
+		ip, err = nodeName2InternalIP(nodeLister, p.ClusterName, instance)
+		if err != nil {
+			return err
+		}
+	}
+	p.Input.MetricParams["instance"] = ip + ".*"
```
Comment (orangedeng) [inline pkg/api/customization/monitor/handler_utils.go]: Is the `instance` also required for not `ResourceNode` type? Is it possible not to set it while the resource type is not `node`?

File: pkg/api/customization/monitor/handler_utils.go [hunk 4]
```diff
@@ -63,40 +65,102 @@ func (p *clusterGraphInputParser) parseFilter() error {
...
+func newClusterAuthChecker(apiContext *types.APIContext, clusterID string) *clusterAuthChecker {
+	return &clusterAuthChecker{
+		clusterID:  clusterID,
+		apiContext: apiContext,
+	}
+}
+
+func (a *clusterAuthChecker) check() error {
+	var cluster mgmtclient.Cluster
+	return access.ByID(a.apiContext, &managementschema.Version, mgmtclient.ClusterType, a.clusterID, &cluster)
```
Comment (orangedeng) [inline pkg/api/customization/monitor/handler_utils.go]: It is not good to check the privileges by accessing to the cluster object. Maybe we need to use something like https://github.com/rancher/rancher/blob/master/pkg/api/customization/cluster/actionhandler.go#L52-L58 to check the privileges.

File: pkg/api/customization/monitor/handler_utils.go [hunk 5]
```diff
@@ -113,34 +177,108 @@ func (p *projectGraphInputParser) parseFilter() error {
...
+		Input:       input,
+		UserContext: userContext,
+		apiContext:  apiContext,
 	}
 }
 
-func (a *authChecker) check() error {
+func (a *projectAuthChecker) check() error {
+	var project mgmtclient.Project
+	if err := access.ByID(a.apiContext, &managementschema.Version, mgmtclient.ProjectType, a.ProjectID, &project); err != nil {
```
Comment (orangedeng) [inline pkg/api/customization/monitor/handler_utils.go]: Same as `clusterAuthChecker`, maybe we need to use something like https://github.com/rancher/rancher/blob/master/pkg/api/customization/cluster/actionhandler.go#L52-L58 to check the privileges.

File: pkg/api/customization/monitor/handler_utils.go (hunk 3 above)
Comment (aiwantaozi) [reply pkg/api/customization/monitor/handler_utils.go]: other resource types including cluster also need to replace the instance with ```.*```

**Comment by orangedeng** [comment]
LGTM

**Comment by prachidamle** [comment]
@aiwantaozi can you please describe with some details the problem fixed - what was missing with the parser and how are we fixing it. This will help me review, Thanks!.

File: pkg/api/customization/monitor/handler_utils.go [hunk 6]
```diff
@@ -63,40 +80,102 @@ func (p *clusterGraphInputParser) parseFilter() error {
 		return fmt.Errorf("clusterId is empty")
```
Comment (prachidamle) [inline pkg/api/customization/monitor/handler_utils.go]: Maybe the error should say something like "Filters is missing, it is a required input" -> message talks about clusterId, but is checking about Filters

File: pkg/api/customization/monitor/handler_utils.go [hunk 7]
```diff
@@ -63,40 +80,102 @@ func (p *clusterGraphInputParser) parseFilter() error {
...
 	}
 
+	p.ResourceType = p.Input.Filters["resourceType"]
+
+	if p.ResourceType == "" {
+		p.ResourceType = p.Input.Filters["displayResourceType"]
+	}
+
+	if p.ResourceType == "" {
+		return fmt.Errorf("resourceType is empty")
```
Comment (prachidamle) [inline pkg/api/customization/monitor/handler_utils.go]: Will be good to specify the exact path to the missing parameter in the error, like Filters["resourceType"]

File: pkg/controllers/user/controllers.go [hunk 8]
```diff
@@ -92,6 +92,7 @@ func RegisterFollower(ctx context.Context, cluster *config.UserContext, kubeConf
 	cluster.RBAC.ClusterRoleBindings("").Controller()
 	cluster.RBAC.RoleBindings("").Controller()
 	cluster.Core.Endpoints("").Controller()
+	workload.NewWorkloadController(ctx, cluster.UserOnlyContext(), nil)
```
Comment (prachidamle) [inline pkg/controllers/user/controllers.go]: Please can you explain why this change is needed?

Comment (aiwantaozi) [reply pkg/controllers/user/controllers.go]: Query graph API in project level need to get workload info, for rancher deployed in HA mode, follower nodes don't have workload info in the cache, so we add it workload here.

**Comment by aiwantaozi** [comment]
updated error message

**Comment by cjellick** [comment]
The description of this PR and commit message carry no real meaning for explaining the change you are making

File: pkg/api/customization/monitor/cluster_graph_action.go [hunk 9]
```diff
@@ -142,12 +150,6 @@ type metricWrap struct {
 }
 
 func graph2Metrics(userContext *config.UserContext, mgmtClient v3.Interface, clusterName, resourceType, refGraphName string, metricSelector, detailsMetricSelector map[string]string, metricParams map[string]string, isDetails bool) ([]*metricWrap, error) {
```
Comment (orangedeng) [inline pkg/api/customization/monitor/cluster_graph_action.go]: I don't see we use `resourceType` parameter anymore. Should it be removed?

Comment (orangedeng) [reply pkg/api/customization/monitor/cluster_graph_action.go]: nvm, I got it.

**Comment by orangedeng** [comment]
LGTM

File: pkg/api/customization/monitor/cluster_graph_action.go [hunk 10]
```diff
@@ -63,6 +68,12 @@ func (h *ClusterGraphHandler) QuerySeriesAction(actionName string, action *types
 		return fmt.Errorf("get usercontext failed, %v", err)
 	}
 
+	mgmtClient := h.clustermanager.ScaledContext.Management
+	nodeLister := mgmtClient.Nodes(metav1.NamespaceAll).Controller().Lister()
```
Comment (cjellick) [inline pkg/api/customization/monitor/cluster_graph_action.go]: never get a lister in business logic like this. you have to get the lister during "setup" of rancher.

**Comment by loganhz** [comment]
will use another solution to fix it
